File: backend/src/main/java/ch/rodano/core/services/bll/user/management/UserManagementService.java
package ch.rodano.core.services.bll.user.management;

import java.util.List;

import ch.rodano.api.user.management.CreateUserRequest;
import ch.rodano.api.user.management.UpdateUserRequest;
import ch.rodano.api.user.management.UserManagementDTO;
import ch.rodano.core.model.audit.DatabaseActionContext;

public interface UserManagementService {

	/**
	 * Get all users (for superuser view)
	 */
	List<UserManagementDTO> getAllUsers();

	/**
	 * Get a single user by pk
	 */
	UserManagementDTO getUser(Long userPk);

	/**
	 * Create a new user
	 */
	UserManagementDTO createUser(CreateUserRequest request, String contextUrl, DatabaseActionContext context);

	/**
	 * Update an existing user
	 */
	UserManagementDTO updateUser(Long userPk, UpdateUserRequest request, DatabaseActionContext context);

	/**
	 * Delete a user (soft delete)
	 */
	void deleteUser(Long userPk, DatabaseActionContext context);

	/**
	 * Restore a deleted user
	 */
	void restoreUser(Long userPk, DatabaseActionContext context);

	/**
	 * Toggle superuser status
	 */
	void toggleSuperuser(Long userPk, boolean isSuperuser, DatabaseActionContext context);

	/**
	 * Unblock a blocked user
	 */
	void unblockUser(Long userPk, DatabaseActionContext context);
}

File: frontends/main/src/app/core/services/user-management.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable, throwError} from 'rxjs';
import {UserManagement} from '@core/model/user-management';
import {CreateUserRequest} from '@core/model/create-user-request';
import {UpdateUserRequest} from '@core/model/update-user-request';

@Injectable({
	providedIn: 'root'
})
export class UserManagementService {
	private readonly baseUrl = '/api/superuser/users';

	constructor(private http: HttpClient) {}

	getAllUsers(): Observable<UserManagement[]> {
		return this.http.get<UserManagement[]>(this.baseUrl);
	}

	getUser(userPk: number | undefined): Observable<UserManagement> {
		if(userPk === undefined) {
			return throwError(() => new Error('User PK is required'));
		}
		return this.http.get<UserManagement>(`${this.baseUrl}/${userPk}`);
	}

	createUser(request: CreateUserRequest): Observable<UserManagement> {
		return this.http.post<UserManagement>(`${this.baseUrl}`, request);
	}

	updateUser(userPk: number | undefined, request: UpdateUserRequest): Observable<UserManagement> {
		if(userPk === undefined) {
			return throwError(() => new Error('User PK is required'));
		}
		return this.http.put(`${this.baseUrl}/${userPk}`, request);
	}

	deleteUser(userPk: number | undefined): Observable<void> {
		if(userPk === undefined) {
			return throwError(() => new Error('User PK is required'));
		}
		return this.http.delete<void>(`${this.baseUrl}/${userPk}`);
	}

	restoreUser(userPk: number | undefined): Observable<void> {
		if(userPk === undefined) {
			return throwError(() => new Error('User PK is required'));
		}
		return this.http.post<void>(`${this.baseUrl}/${userPk}/restore`, {});
	}

	toggleSuperuser(userPk: number | undefined, isSuperuser: boolean): Observable<void> {
		if(userPk === undefined) {
			return throwError(() => new Error('User PK is required'));
		}
		return this.http.put<void>(`${this.baseUrl}/${userPk}/superuser`, {isSuperuser});
	}

	unblockUser(userPk: number | undefined): Observable<void> {
		if(userPk === undefined) {
			return throwError(() => new Error('User PK is required'));
		}
		return this.http.post<void>(`${this.baseUrl}/${userPk}/unblock`, {});
	}
}
